File: src/quadropted_controller_cpp/src/nodes/robot_controller_node.cpp
#include "quadropted_controller_cpp/nodes/robot_controller_node.hpp"

#include <algorithm>
#include <cmath>
#include <memory>
#include <string>

namespace quadropted {

RobotControllerNode::RobotControllerNode() : Node("robot_controller_cpp"), state_(0.25) {
    declare_parameter("rate", 60);
    declare_parameter("body_length", 0.3762);
    declare_parameter("body_width", 0.0935);
    declare_parameter("leg_l1", 0.0);
    declare_parameter("leg_l2", 0.0955);
    declare_parameter("leg_l3", 0.213);
    declare_parameter("leg_l4", 0.213);
    declare_parameter("dx_front_offset", 0.02);
    declare_parameter("dx_back_offset", 0.0);
    declare_parameter("trot_stance_time", 0.04);
    declare_parameter("trot_swing_time", 0.18);
    declare_parameter("trot_dt", 0.02);
    declare_parameter("trot_use_imu", false);
    declare_parameter("trot_z_leg_lift", 0.14);
    declare_parameter("trot_z_error_constant", 0.02);
    declare_parameter("trot_pid_kp", 0.15);
    declare_parameter("trot_pid_ki", 0.02);
    declare_parameter("trot_pid_kd", 0.002);
    declare_parameter("crawl_stance_time", 0.55);
    declare_parameter("crawl_swing_time", 0.45);
    declare_parameter("crawl_dt", 0.02);
    declare_parameter("crawl_z_leg_lift", 0.14);
    declare_parameter("crawl_body_shift_y", 0.06);
    declare_parameter("crawl_z_error_constant", 0.02);
    declare_parameter("crawl_max_vx", 0.011);
    declare_parameter("crawl_max_vy_scale", 0.5);
    declare_parameter("crawl_max_yaw", 0.15);
    declare_parameter("rest_pid_kp", 0.75);
    declare_parameter("rest_pid_ki", 2.29);
    declare_parameter("rest_pid_kd", 0.0);
    declare_parameter("stand_body_velocity_scale", 0.01);
    declare_parameter("stand_body_angular_scale", 0.005);
    declare_parameter("stand_max_linear_velocity", 0.2);
    declare_parameter("stand_max_angular_velocity", 0.5);
    declare_parameter("sit_height", -0.15);
    declare_parameter("stand_height", 0.005);
    declare_parameter("walk_height", 0.0);
    declare_parameter("startup_grace_ticks", 120);
    declare_parameter("verbose", false);

    rate_ = get_parameter("rate").as_int();
    double body_length = get_parameter("body_length").as_double();
    double body_width = get_parameter("body_width").as_double();
    double leg_l1 = get_parameter("leg_l1").as_double();
    double leg_l2 = get_parameter("leg_l2").as_double();
    double leg_l3 = get_parameter("leg_l3").as_double();
    double leg_l4 = get_parameter("leg_l4").as_double();
    double dx_front_offset = get_parameter("dx_front_offset").as_double();
    double dx_back_offset = get_parameter("dx_back_offset").as_double();
    double trot_stance_time = get_parameter("trot_stance_time").as_double();
    double trot_swing_time = get_parameter("trot_swing_time").as_double();
    double trot_dt = get_parameter("trot_dt").as_double();
    bool trot_use_imu = get_parameter("trot_use_imu").as_bool();
    double trot_z_leg_lift = get_parameter("trot_z_leg_lift").as_double();
    double trot_z_error_constant = get_parameter("trot_z_error_constant").as_double();
    double trot_pid_kp = get_parameter("trot_pid_kp").as_double();
    double trot_pid_ki = get_parameter("trot_pid_ki").as_double();
    double trot_pid_kd = get_parameter("trot_pid_kd").as_double();
    double crawl_stance_time = get_parameter("crawl_stance_time").as_double();
    double crawl_swing_time = get_parameter("crawl_swing_time").as_double();
    double crawl_dt = get_parameter("crawl_dt").as_double();
    double crawl_z_leg_lift = get_parameter("crawl_z_leg_lift").as_double();
    double crawl_body_shift_y = get_parameter("crawl_body_shift_y").as_double();
    double crawl_z_error_constant = get_parameter("crawl_z_error_constant").as_double();
    double crawl_max_vx = get_parameter("crawl_max_vx").as_double();
    double crawl_max_vy_scale = get_parameter("crawl_max_vy_scale").as_double();
    double crawl_max_yaw = get_parameter("crawl_max_yaw").as_double();
    double rest_pid_kp = get_parameter("rest_pid_kp").as_double();
    double rest_pid_ki = get_parameter("rest_pid_ki").as_double();
    double rest_pid_kd = get_parameter("rest_pid_kd").as_double();
    double stand_body_velocity_scale = get_parameter("stand_body_velocity_scale").as_double();
    double stand_body_angular_scale = get_parameter("stand_body_angular_scale").as_double();
    double stand_max_linear_velocity = get_parameter("stand_max_linear_velocity").as_double();
    double stand_max_angular_velocity = get_parameter("stand_max_angular_velocity").as_double();
    sit_height_ = get_parameter("sit_height").as_double();
    stand_height_ = get_parameter("stand_height").as_double();
    walk_height_ = get_parameter("walk_height").as_double();
    startup_grace_ = get_parameter("startup_grace_ticks").as_int();
    debug_mode_ = get_parameter("verbose").as_bool();

    double dx_front = body_length * 0.5 + dx_front_offset;
    double dx_back = body_length * 0.5 + dx_back_offset;
    double dy = body_width * 0.5 + leg_l2;
    default_stance_ << dx_front, dx_front, -dx_back, -dx_back, -dy, dy, -dy, dy, 0, 0, 0, 0;

    state_.foot_locations = default_stance_;
    state_.behavior_state = BehaviorState::REST;

    trot_gait_ = std::make_unique<TrotGaitController>(trot_stance_time, trot_swing_time, trot_dt, trot_use_imu,
                                                       default_stance_, trot_z_leg_lift, trot_z_error_constant,
                                                       trot_pid_kp, trot_pid_ki, trot_pid_kd);
    crawl_gait_ = std::make_unique<CrawlGaitController>(crawl_stance_time, crawl_swing_time, crawl_dt, default_stance_,
                                                        crawl_z_leg_lift, crawl_body_shift_y, crawl_z_error_constant);
    rest_ctrl_ = std::make_unique<RestController>(default_stance_, rest_pid_kp, rest_pid_ki, rest_pid_kd);
    stand_ctrl_ = std::make_unique<StandController>(default_stance_, stand_body_velocity_scale, stand_body_angular_scale,
                                                    stand_max_linear_velocity, stand_max_angular_velocity);

    command_.trot_event = true;
    command_.rest_event = true;
    change_controller();

    ik_ = std::make_unique<InverseKinematics>(body_length, body_width, leg_l1, leg_l2, leg_l3, leg_l4);

    joint_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>("joint_group_controller/commands", 10);
    foot_contact_pub_ =
        create_publisher<quadropted_msgs::msg::RobotFootContact>("foot_contact", rclcpp::SensorDataQoS());

    velocity_sub_ = create_subscription<quadropted_msgs::msg::RobotVelocity>(
        "robot_velocity", 10,
        [this, crawl_max_vx, crawl_max_vy_scale,
         crawl_max_yaw](const quadropted_msgs::msg::RobotVelocity::SharedPtr msg) {
            if (msg->robot_id == 1) {
                command_.velocity = {msg->cmd_vel.linear.x, msg->cmd_vel.linear.y, msg->cmd_vel.linear.z};
                command_.yaw_rate = {msg->cmd_vel.angular.x, msg->cmd_vel.angular.y, msg->cmd_vel.angular.z};

                if (state_.behavior_state == BehaviorState::STAND) {
                    RCLCPP_INFO(get_logger(), "[STAND VELOCITY] vx=%.4f vy=%.4f vz=%.4f | ax=%.4f ay=%.4f az=%.4f",
                                command_.velocity[0], command_.velocity[1], command_.velocity[2], command_.yaw_rate[0],
                                command_.yaw_rate[1], command_.yaw_rate[2]);
                }

                if (state_.behavior_state == BehaviorState::CRAWL) {
                    command_.velocity[0] = std::clamp(command_.velocity[0], -crawl_max_vx, crawl_max_vx);
                    command_.velocity[1] = std::clamp(command_.velocity[1], -crawl_max_vx * crawl_max_vy_scale,
                                                      crawl_max_vx * crawl_max_vy_scale);
                    command_.yaw_rate[2] = std::clamp(command_.yaw_rate[2], -crawl_max_yaw, crawl_max_yaw);
                }
            }
        });

    imu_sub_ =
        create_subscription<sensor_msgs::msg::Imu>("imu", 10, [this](const sensor_msgs::msg::Imu::SharedPtr msg) {
            double w = msg->orientation.w;
            double x = msg->orientation.x;
            double y = msg->orientation.y;
            double z = msg->orientation.z;
            state_.imu_roll = std::atan2(2.0 * (w * x + y * z), 1.0 - 2.0 * (x * x + y * y));
            state_.imu_pitch = std::asin(2.0 * (w * y - z * x));
        });

    mode_sub_ = create_subscription<quadropted_msgs::msg::RobotModeCommand>(
        "robot_mode", 10, [this](const quadropted_msgs::msg::RobotModeCommand::SharedPtr msg) {
            if (msg->robot_id == 1) {
                if (msg->mode == "REST") {
                    command_.rest_event = true;
                    command_.trot_event = false;
                    command_.crawl_event = false;
                    command_.stand_event = false;
                } else if (msg->mode == "TROT") {
                    command_.rest_event = false;
                    command_.trot_event = true;
                    command_.crawl_event = false;
                    command_.stand_event = false;
                } else if (msg->mode == "CRAWL") {
                    command_.rest_event = false;
                    command_.trot_event = false;
                    command_.crawl_event = true;
                    command_.stand_event = false;
                } else if (msg->mode == "STAND") {
                    command_.rest_event = false;
                    command_.trot_event = false;
                    command_.crawl_event = false;
                    command_.stand_event = true;
                }
                change_controller();
            }
        });

    timer_ = create_wall_timer(std::chrono::microseconds(static_cast<long long>(1000000.0 / rate_)),
                               std::bind(&RobotControllerNode::control_loop, this));

    RCLCPP_INFO(get_logger(), "Robot Controller Node (C++) started at %d Hz", rate_);
    RCLCPP_INFO(get_logger(), "Startup grace period: 2 seconds (waiting for robot to land)");

    behavior_srv_ = create_service<quadropted_msgs::srv::RobotBehaviorCommand>(
        "robot_behavior_command",
        [this](const std::shared_ptr<quadropted_msgs::srv::RobotBehaviorCommand::Request> request,
               std::shared_ptr<quadropted_msgs::srv::RobotBehaviorCommand::Response> response) {
            std::string cmd = request->command;
            std::transform(cmd.begin(), cmd.end(), cmd.begin(), ::tolower);
            RCLCPP_INFO(get_logger(), "Received behavior command: %s", cmd.c_str());

            if (cmd == "sit") {
                command_.stand_event = true;
                command_.rest_event = false;
                command_.trot_event = false;
                command_.crawl_event = false;
                change_controller();
                state_.body_local_position[2] = sit_height_;
                response->success = true;
                response->message = "Robot sat down.";
            } else if (cmd == "up") {
                command_.rest_event = true;
                command_.stand_event = false;
                command_.trot_event = false;
                command_.crawl_event = false;
                change_controller();
                state_.body_local_position[2] = stand_height_;
                response->success = true;
                response->message = "Robot stood up.";
            } else if (cmd == "walk") {
                command_.rest_event = true;
                command_.trot_event = true;
                command_.stand_event = false;
                command_.crawl_event = false;
                change_controller();
                state_.body_local_position[2] = walk_height_;
                response->success = true;
                response->message = "Robot started walking.";
            } else {
                response->success = false;
                response->message = "Unknown command: " + request->command;
            }
        });
}

void RobotControllerNode::change_controller() {
    if (command_.trot_event && command_.rest_event) {
        state_.behavior_state = BehaviorState::REST;
        rest_ctrl_->pid().reset(this->now().seconds());
        command_.rest_event = false;

        state_.behavior_state = BehaviorState::TROT;
        trot_gait_->pid_controller().reset(this->now().seconds());
        state_.ticks = 0;
        state_.body_local_position[2] = walk_height_;
        command_.trot_event = false;
        RCLCPP_INFO(get_logger(), "Switched to TROT controller");
    } else if (command_.trot_event) {
        auto prev = state_.behavior_state;
        state_.behavior_state = BehaviorState::TROT;
        trot_gait_->pid_controller().reset(this->now().seconds());
        state_.ticks = 0;
        state_.body_local_position[2] = walk_height_;
        command_.trot_event = false;
        if (prev == BehaviorState::CRAWL) {
            RCLCPP_INFO(get_logger(), "Switched to TROT controller (from CRAWL)");
        } else if (prev == BehaviorState::STAND) {
            RCLCPP_INFO(get_logger(), "Switched to TROT controller (from STAND)");
        }
    } else if (command_.rest_event) {
        state_.behavior_state = BehaviorState::REST;
        rest_ctrl_->pid().reset(this->now().seconds());
        state_.body_local_position[2] = sit_height_;
        command_.rest_event = false;
        RCLCPP_INFO(get_logger(), "Switched to REST controller — lying down");
    } else if (command_.stand_event) {
        if (state_.behavior_state != BehaviorState::STAND) {
            state_.behavior_state = BehaviorState::STAND;
            state_.body_local_position[2] = stand_height_;
            RCLCPP_INFO(get_logger(), "Switched to STAND controller");
        }
        command_.stand_event = false;
    } else if (command_.crawl_event) {
        state_.behavior_state = BehaviorState::CRAWL;
        crawl_gait_->reset();
        state_.ticks = 0;
        state_.body_local_position[2] = walk_height_;
        command_.crawl_event = false;
    }
}

void RobotControllerNode::publish_foot_contacts() {
    auto msg = std::make_unique<quadropted_msgs::msg::RobotFootContact>();
    switch (state_.behavior_state) {
        case BehaviorState::REST:
        case BehaviorState::STAND:
            msg->contacts = {true, true, true, true};
            break;
        case BehaviorState::TROT: {
            Eigen::VectorXi contacts = trot_gait_->contacts(state_.ticks);
            msg->contacts = {contacts(0) != 0, contacts(1) != 0, contacts(2) != 0, contacts(3) != 0};
            break;
        }
        case BehaviorState::CRAWL: {
            Eigen::VectorXi contacts = crawl_gait_->contacts(state_.ticks);
            msg->contacts = {contacts(0) != 0, contacts(1) != 0, contacts(2) != 0, contacts(3) != 0};
            break;
        }
    }
    foot_contact_pub_->publish(std::move(msg));
}

void RobotControllerNode::control_loop() {
    if (startup_grace_ > 0) {
        startup_grace_--;
        if (startup_grace_ == 0) {
            RCLCPP_INFO(get_logger(), "Startup grace period complete, controller active");
        }
        return;
    }

    rclcpp::Time now = this->now();
    LegsMatrix leg_positions;
    switch (state_.behavior_state) {
        case BehaviorState::TROT:
            leg_positions = step_trot(state_, command_, now.seconds());
            break;
        case BehaviorState::CRAWL:
            leg_positions = step_crawl(state_, command_);
            break;
        case BehaviorState::STAND:
            leg_positions = step_stand(state_, command_);
            break;
        case BehaviorState::REST:
        default:
            leg_positions = step_rest(state_, command_);
            break;
    }

    state_.foot_locations = leg_positions;
    state_.robot_height = command_.robot_height;

    publish_foot_contacts();

    if (state_.ticks < 5) {
        RCLCPP_INFO(get_logger(), "[IK DEBUG] leg_positions: %dx%d, dx=%.3f dy=%.3f dz=%.3f", (int)leg_positions.rows(),
                    (int)leg_positions.cols(), state_.body_local_position[0], state_.body_local_position[1],
                    state_.robot_height);
    }

    try {
        auto joint_angles = ik_->inverse_kinematics(
            leg_positions, state_.body_local_position[0], state_.body_local_position[1], state_.body_local_position[2],
            state_.body_local_orientation[0], state_.body_local_orientation[1], state_.body_local_orientation[2]);

        auto msg = std::make_unique<std_msgs::msg::Float64MultiArray>();
        msg->data.assign(joint_angles.begin(), joint_angles.end());
        joint_pub_->publish(std::move(msg));

        if (state_.ticks % 60 == 0) {
            RCLCPP_INFO(get_logger(),
                        "[DEBUG] cmd: vx=%.4f vy=%.4f vz=%.4f yaw=%.4f | "
                        "pos: x=%.4f y=%.4f z=%.4f | "
                        "joints[0-2]: %.4f %.4f %.4f",
                        command_.velocity[0], command_.velocity[1], command_.velocity[2], command_.yaw_rate[2],
                        state_.body_local_position[0], state_.body_local_position[1], state_.body_local_position[2],
                        joint_angles[0], joint_angles[1], joint_angles[2]);
        }
    } catch (const std::exception& e) {
        RCLCPP_ERROR(get_logger(), "IK error: %s", e.what());
    }
}

}  // namespace quadropted

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<quadropted::RobotControllerNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/quadropted_controller_rust/quadropted-nodes/src/bin/robot_controller_node.rs
//! Robot Controller Node — Full Rust implementation with State Machine
//!
//! Full ROS 2 integration:
//! - Behavior State Machine: REST/TROT/CRAWL/STAND
//! - Subscriptions: robot_mode, robot_velocity, imu
//! - TrotGaitController + CrawlGaitController
//! - IK computes joint angles from foot positions
//! - Float64MultiArray publisher for joint_group_controller/commands

use std_msgs_rs::Float64MultiArray;
use quadropted_core::controllers::trot::gait::TrotGaitController;
use quadropted_core::controllers::crawl::gait::CrawlGaitController;
use quadropted_core::controllers::rest::{RestController, RestState};
use quadropted_core::controllers::stand::{StandController, BodyState};
use quadropted_core::state::behavior::BehaviorState;
use quadropted_core::kinematics::inverse::{compute_local_positions, compute_all_joint_angles};
use nalgebra::SMatrix;
use rclrs::{Context, CreateBasicExecutor, Publisher, SpinOptions};
use rosidl_runtime_rs::Sequence;
use std::sync::{Arc, Mutex};
use std::time::Duration;

struct SharedState {
    ticks: i32,
    foot_locations: SMatrix<f64, 3, 4>,
    default_stance: SMatrix<f64, 3, 4>,
    behavior_state: BehaviorState,
    trot_gait: TrotGaitController,
    crawl_gait: CrawlGaitController,
    rest_ctrl: RestController,
    rest_state: RestState,
    stand_ctrl: StandController,
    body_state: BodyState,
    cmd_linear: [f64; 3],
    cmd_angular: [f64; 3],
    imu_roll: f64,
    imu_pitch: f64,
    mode_msg_count: u64,
    vel_msg_count: u64,
    startup_grace: i32,
}

impl SharedState {
    fn new() -> Self {
        let body_length = 0.3762;
        let body_width = 0.0935;
        let l2 = 0.0955;
        let dx_front = body_length * 0.5 + 0.02;
        let dx_back = body_length * 0.5;
        let dy = body_width * 0.5 + l2;

        let mut default_stance = SMatrix::<f64, 3, 4>::zeros();
        default_stance[(0, 0)] = dx_front; default_stance[(1, 0)] = -dy;
        default_stance[(0, 1)] = dx_front; default_stance[(1, 1)] = dy;
        default_stance[(0, 2)] = -dx_back; default_stance[(1, 2)] = -dy;
        default_stance[(0, 3)] = -dx_back; default_stance[(1, 3)] = dy;

        let trot_gait = TrotGaitController::new(0.04, 0.18, 0.02, false, default_stance.clone());
        let crawl_gait = CrawlGaitController::new(0.55, 0.45, 0.02, default_stance.clone());
        let rest_ctrl = RestController::new(default_stance.clone());
        let stand_ctrl = StandController::new(default_stance.clone());

        println!("[Rust] Default stance:");
        for leg in 0..4 {
            let name = ["FR", "FL", "RR", "RL"][leg];
            println!("[Rust]   {}: x={:.4} y={:.4}", name, default_stance[(0, leg)], default_stance[(1, leg)]);
        }

        Self {
            ticks: 0,
            foot_locations: default_stance.clone(),
            default_stance,
            behavior_state: BehaviorState::REST,
            trot_gait,
            crawl_gait,
            rest_ctrl,
            rest_state: RestState { imu_roll: 0.0, imu_pitch: 0.0 },
            stand_ctrl,
            body_state: BodyState {
                body_local_position: [0.0, 0.0, 0.0],
                body_local_orientation: [0.0, 0.0, 0.0],
            },
            cmd_linear: [0.0, 0.0, 0.0],
            cmd_angular: [0.0, 0.0, 0.0],
            imu_roll: 0.0,
            imu_pitch: 0.0,
            mode_msg_count: 0,
            vel_msg_count: 0,
            startup_grace: 120, // 2 сек @ 60 Гц (как C++ startup_grace_)
        }
    }

    fn step(&mut self, robot_height: f64) -> [f64; 12] {
        self.ticks += 1;

        // State machine: select controller based on behavior_state
        self.foot_locations = match self.behavior_state {
            BehaviorState::REST => {
                self.rest_ctrl.step(&self.rest_state, robot_height)
            }
            BehaviorState::TROT => {
                let gait_cmd = [self.cmd_linear[0], self.cmd_linear[1], self.cmd_angular[2]];
                // C++ step_trot: при нулевой скорости — плавное возвращение к default_stance
                let has_command =
                    gait_cmd[0].abs() > 1e-4 || gait_cmd[1].abs() > 1e-4 || gait_cmd[2].abs() > 1e-4;
                if !has_command {
                    let mut result = self.default_stance;
                    result.row_mut(2).fill(robot_height);
                    let alpha = 0.1;
                    self.foot_locations * (1.0 - alpha) + result * alpha
                } else {
                    let mut new_foot = self.trot_gait.step(
                        self.ticks,
                        &self.foot_locations,
                        &gait_cmd,
                        robot_height,
                    );
                    // IMU compensation (как в C++ step_trot)
                    if self.trot_gait.use_imu() {
                        let now_sec = std::time::SystemTime::now()
                            .duration_since(std::time::UNIX_EPOCH)
                            .map(|d| d.as_secs_f64())
                            .unwrap_or(0.0);
                        let comp = self.trot_gait.pid_controller().run(self.imu_roll, self.imu_pitch, now_sec);
                        let rot = quadropted_core::math::rotation::rotxyz(-comp[0], -comp[1], 0.0);
                        new_foot = rot * new_foot;
                    }
                    new_foot
                }
            }
            BehaviorState::CRAWL => {
                // Clamp velocity for crawl mode
                let mut crawl_vel = [self.cmd_linear[0], self.cmd_linear[1], self.cmd_angular[2]];
                crawl_vel[0] = crawl_vel[0].clamp(-0.011, 0.011);
                crawl_vel[1] = crawl_vel[1].clamp(-0.0055, 0.0055);
                crawl_vel[2] = crawl_vel[2].clamp(-0.15, 0.15);

                self.crawl_gait.step(self.ticks, &self.foot_locations, &crawl_vel, robot_height)
            }
            BehaviorState::STAND => {
                self.stand_ctrl.run(
                    &mut self.body_state,
                    robot_height,
                    &self.cmd_linear,
                    &self.cmd_angular,
                )
            }
        };

        // IK: foot positions → joint angles
        // C++ передаёт body_local_position/orientation (высота тела из change_controller)
        let bp = &self.body_state.body_local_position;
        let bo = &self.body_state.body_local_orientation;
        let local = compute_local_positions(
            &self.foot_locations, 0.3762, 0.0935,
            bp[0], bp[1], bp[2], bo[0], bo[1], bo[2],
        );
        let angles = compute_all_joint_angles(&local, 0.0, 0.0955, 0.213, 0.213);

        angles
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🦀 Rust Robot Controller Node starting...");
    println!("   Features: State Machine (REST/TROT/CRAWL/STAND) + IK + Subscriptions\n");

    // Парсим аргументы процесса (--ros-args -r __ns:=/robot1 от launch), иначе
    // namespace и remappings из launch-файла не применяются (rclrs 0.7).
    let ctx = Context::new(std::env::args(), rclrs::InitOptions::new())?;
    let mut executor = ctx.create_basic_executor();
    let node = executor.create_node("robot_controller_rust")?;
    println!("✅ Node created: robot_controller_rust");

    let state = Arc::new(Mutex::new(SharedState::new()));

    // Publisher for joint commands
    let joint_pub: Publisher<Float64MultiArray> =
        node.create_publisher("joint_group_controller/commands")?;
    println!("✅ Publisher: joint_group_controller/commands");

    // Publisher for foot contacts (для odometry — как C++ publish_foot_contacts)
    let contact_pub: Publisher<quadropted_msgs_rs::RobotFootContact> =
        node.create_publisher("foot_contact")?;
    println!("✅ Publisher: foot_contact");

    // Subscription: robot_mode
    let mode_state = state.clone();
    let _mode_sub = node.create_subscription("robot_mode", move |msg: quadropted_msgs_rs::RobotModeCommand| {
        if msg.robot_id == 1 {
            let mut s = mode_state.lock().unwrap();
            let mode_raw = msg.mode.to_cstr().to_string_lossy();
            let mode_str = mode_raw.trim().trim_matches(char::from(0));
            if let Some(new_state) = BehaviorState::from_str(mode_str) {
                s.mode_msg_count += 1;
                println!("[Rust] Mode change: {:?} -> {:?}", s.behavior_state, new_state);
                s.behavior_state = new_state;
                s.ticks = 0;

                // Reset controllers on mode change (как в C++ change_controller)
                match new_state {
                    BehaviorState::CRAWL => {
                        s.crawl_gait.reset();
                        s.body_state.body_local_position[2] = 0.0;
                    }
                    BehaviorState::TROT => {
                        // C++: trot_gait_->pid_controller().reset(this->now().seconds())
                        s.trot_gait.pid_controller().reset(
                            std::time::SystemTime::now()
                                .duration_since(std::time::UNIX_EPOCH)
                                .map(|d| d.as_secs_f64())
                                .unwrap_or(0.0),
                        );
                        s.body_state.body_local_position[2] = 0.0;
                    }
                    BehaviorState::REST => {
                        // C++: body_local_position[2] = -0.15 (лечь на землю)
                        s.body_state.body_local_position[2] = -0.15;
                    }
                    BehaviorState::STAND => {
                        // C++: body_local_position[2] = 0.005
                        s.body_state.body_local_position[2] = 0.005;
                    }
                }
            } else {
                println!("[Rust] Ignored unknown mode: '{}'", mode_str);
            }
        }
    })?;
    println!("✅ Subscription: robot_mode");

    // Subscription: robot_velocity
    let vel_state = state.clone();
    let _vel_sub = node.create_subscription("robot_velocity", move |msg: quadropted_msgs_rs::RobotVelocity| {
        if msg.robot_id == 1 {
            let mut s = vel_state.lock().unwrap();
            s.vel_msg_count += 1;
            s.cmd_linear = [
                msg.cmd_vel.linear.x,
                msg.cmd_vel.linear.y,
                msg.cmd_vel.linear.z,
            ];
            s.cmd_angular = [
                msg.cmd_vel.angular.x,
                msg.cmd_vel.angular.y,
                msg.cmd_vel.angular.z,
            ];
        }
    })?;
    println!("✅ Subscription: robot_velocity");

    // Subscription: imu
    let imu_state = state.clone();
    let _imu_sub = node.create_subscription("imu", move |msg: sensor_msgs_rs::Imu| {
        let mut s = imu_state.lock().unwrap();
        // Convert quaternion to euler angles
        let w = msg.orientation.w;
        let x = msg.orientation.x;
        let y = msg.orientation.y;
        let z = msg.orientation.z;
        s.imu_roll = (2.0 * (w * x + y * z)).atan2(1.0 - 2.0 * (x * x + y * y));
        s.imu_pitch = (2.0 * (w * y - z * x)).asin();
    })?;
    println!("✅ Subscription: imu");

    // Service: robot_behavior_command (sit/up/walk) — как C++ behavior_srv_
    let srv_state = state.clone();
    let _behavior_srv = node.create_service::<quadropted_msgs_rs::RobotBehaviorCommand, _>(
        "robot_behavior_command",
        move |req: quadropted_msgs_rs::RobotBehaviorCommand_Request| {
            let mut resp = quadropted_msgs_rs::RobotBehaviorCommand_Response::default();
            let cmd = req.command.to_cstr().to_string_lossy().to_lowercase();
            println!("[Rust] Received behavior command: {}", cmd);

            let mut s = srv_state.lock().unwrap();
            match cmd.as_str() {
                "sit" => {
                    s.behavior_state = BehaviorState::STAND;
                    s.ticks = 0;
                    s.body_state.body_local_position[2] = -0.15;
                    resp.success = true;
                    resp.message = "Robot sat down.".into();
                }
                "up" => {
                    s.behavior_state = BehaviorState::REST;
                    s.ticks = 0;
                    s.body_state.body_local_position[2] = 0.0;
                    resp.success = true;
                    resp.message = "Robot stood up.".into();
                }
                "walk" => {
                    s.behavior_state = BehaviorState::REST;
                    s.ticks = 0;
                    s.body_state.body_local_position[2] = 0.0;
                    // C++: rest_event + trot_event → REST затем TROT
                    s.behavior_state = BehaviorState::TROT;
                    s.ticks = 0;
                    s.trot_gait.pid_controller().reset(
                        std::time::SystemTime::now()
                            .duration_since(std::time::UNIX_EPOCH)
                            .map(|d| d.as_secs_f64())
                            .unwrap_or(0.0),
                    );
                    resp.success = true;
                    resp.message = "Robot started walking.".into();
                }
                other => {
                    resp.success = false;
                    resp.message = format!("Unknown command: {}", other).into();
                }
            }
            resp
        },
    )?;
    println!("✅ Service: robot_behavior_command");

    // 60Hz control loop
    let ctrl_state = state.clone();
    let ctrl_pub = joint_pub.clone();
    std::thread::spawn(move || loop {
        let mut s = ctrl_state.lock().unwrap();

        // Startup grace period: ждём пока робот приземлится (как C++ startup_grace_)
        if s.startup_grace > 0 {
            s.startup_grace -= 1;
            if s.startup_grace == 0 {
                println!("[Rust] Startup grace period complete, controller active");
            }
            drop(s);
            std::thread::sleep(Duration::from_millis(16));
            continue;
        }

        let angles = s.step(-0.25);

        if s.ticks % 120 == 0 {
            println!("[Rust DEBUG] Tick #{} ({:.1}s) {:?} mode, vx={:.3}",
                s.ticks, s.ticks as f64 / 60.0, s.behavior_state, s.cmd_linear[0]);
        }

        let mut msg = Float64MultiArray::default();
        let mut seq = Sequence::new(angles.len());
        for (i, &val) in angles.iter().enumerate() { seq[i] = val; }
        msg.data = seq;
        ctrl_pub.publish(&msg).ok();

        // Публикация контактов ног (для odometry) — как C++ publish_foot_contacts
        let contact_msg = {
            let mut cm = quadropted_msgs_rs::RobotFootContact::default();
            match s.behavior_state {
                BehaviorState::REST | BehaviorState::STAND => {
                    cm.contacts = [true, true, true, true];
                }
                BehaviorState::TROT => {
                    let c = s.trot_gait.contacts(s.ticks);
                    cm.contacts = [c[0] != 0, c[1] != 0, c[2] != 0, c[3] != 0];
                }
                BehaviorState::CRAWL => {
                    let c = s.crawl_gait.contacts(s.ticks);
                    cm.contacts = [c[0] != 0, c[1] != 0, c[2] != 0, c[3] != 0];
                }
            }
            cm
        };
        contact_pub.publish(&contact_msg).ok();

        drop(s);
        std::thread::sleep(Duration::from_millis(16)); // 60Hz
    });

    println!("✅ 60Hz control loop with State Machine");
    println!("🚀 Spinning (Ctrl+C to stop)...\n");

    loop {
        let mut spin_opts = SpinOptions::default();
        spin_opts.timeout = Some(Duration::from_millis(1000));
        let _ = executor.spin(spin_opts);
    }
}
